src/array.jl

```diff
@@ -482,27 +482,58 @@ end
 
 # optimize reshape to return a CuArray
 
-function Base.reinterpret(::Type{T}, a::CuArray{S,N}) where {T,S,N}
-  if N == 0 && sizeof(T) == sizeof(S)
-    throw(ArgumentError("cannot reinterpret a zero-dimensional `$(S)` array to `$(T)` which is of a different size"))
+struct _CuReinterpretZeroDimError{T,A} <: Exception end
+function Base.showerror(io::IO, ::_CuReinterpretZeroDimError{T, <:AbstractArray{S,N}}) where {T, S, N}
+  print(io, "cannot reinterpret a zero-dimensional `$(S)` array to `$(T)` which is of a different size")
+end
```

Comment:
Do we need this approach for constructing an error message for an exception that could be thrown on the device?

(And I guess something like this is what you want to automate via Opaque Closure?)

Reply:
The problem with exceptions is that they are often `@nospecialize`, so lead to allocating, GPU-incompatible code. That's why GPUCompiler.jl has some heuristics to catch the throwing function, which is nowadays often outlined to avoid a GC frame in hot code (e.g. `throw_boundserror`). Would be nice to automate this and ensure we can catch the throwing function, even though it is a workaround (disabling `@nospecialize` or allowing allocations would be a "proper" fix).

That said, if you just `error` you don't need this on-device, but you also won't get to see the actual error message.

Reply:
Do we need the error types? Why not just outline and `error` or `throw(ArgumentError(...))`?

Reply:
I thought to create the exception types so that the device code won't try to construct the error message strings. Code like this `Base.rem(dim*sizeof(S),sizeof(T)) != 0` requires run-time check so I guess the compiler can't optimize out some error pass. Without the exception types, I saw a compilation error `unsupported dynamic function invocation (call to print_to_string(xs...) in Base at strings/io.jl:124` presumably due to string interpolation. Or can this be done with error-throwing `@noinline` functions?

Reply:
Yeah, but it's not particularly nice: https://github.com/JuliaGPU/GPUCompiler.jl/blob/ff66447286f6646c0d796487b073b767488dc4a1/src/irgen.jl#L176-L198. That should catch calls to `throw` and `error` though.

Reply:
Hmm... How do I use `@noinline` to throw an error on kernel? I tried

```julia
julia> function k1()
           threadIdx().x > 1 && throw(Some(threadIdx().x))
           return
       end
k1 (generic function with 1 method)

julia> CUDA.@sync @cuda threads=1 blocks=1 k1()
CUDA.HostKernel{k1, Tuple{}}(CuContext(0x00000000038426d0, instance 6173ea0f45abefbe), CuModule(Ptr{Nothing} @0x0000000007083f50, CuContext(0x00000000038426d0, instance 6173ea0f45abefbe)), CuFunction(Ptr{Nothing} @0x0000000007373300, CuModule(Ptr{Nothing} @0x0000000007083f50, CuContext(0x00000000038426d0, instance 6173ea0f45abefbe))))

julia> @noinline f(i) = error("i=$i")
f (generic function with 1 method)

julia> function k2()
           threadIdx().x > 1 && f(threadIdx().x)
           return
       end
k2 (generic function with 1 method)

julia> CUDA.@sync @cuda threads=1 blocks=1 k2()
ERROR: InvalidIRError: compiling kernel k2() resulted in invalid LLVM IR
Reason: unsupported dynamic function invocation (call to print_to_string(xs...) in Base at strings/io.jl:124)
Stacktrace:
 [1] string
   @ strings/io.jl:174
 [2] f
   @ REPL[13]:1
 [3] k2
   @ REPL[14]:2
Stacktrace:
  [1] check_ir(job::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams, GPUCompiler.FunctionSpec{typeof(k2), Tuple{}}}, args::LLVM.Module)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/XwWPj/src/validation.jl:123
```

`k1` is equivalent to what am I doing here and it "works" as long as it hits the happy path. But I couldn't make something like `k2` compilable.

Reply:
Oh hmm, we have some code that kills of arguments to exceptions (such as the formatted string here being passed to `ErrorException` by `error(...)`), https://github.com/JuliaGPU/GPUCompiler.jl/blob/ff66447286f6646c0d796487b073b767488dc4a1/src/irgen.jl#L220-L232, but apparently that doesn't work in this case.

Reply:
It should already work with plain `throw`:

```julia
julia> kernel(x) = (x && throw("$x"); nothing)
kernel (generic function with 1 method)

julia> kernel(x) = (x ? throw("$x"); nothing)^C

julia> @cuda kernel(false)
CUDA.HostKernel{kernel, Tuple{Bool}}(CuContext(0x00000000026b2bc0, instance fe73669339970426), CuModule(Ptr{Nothing} @0x00000000072844f0, CuContext(0x00000000026b2bc0, instance fe73669339970426)), CuFunction(Ptr{Nothing} @0x00000000073ea910, CuModule(Ptr{Nothing} @0x00000000072844f0, CuContext(0x00000000026b2bc0, instance fe73669339970426))))

julia> @cuda kernel(true)
CUDA.HostKernel{kernel, Tuple{Bool}}(CuContext(0x00000000026b2bc0, instance fe73669339970426), CuModule(Ptr{Nothing} @0x00000000072844f0, CuContext(0x00000000026b2bc0, instance fe73669339970426)), CuFunction(Ptr{Nothing} @0x00000000073ea910, CuModule(Ptr{Nothing} @0x00000000072844f0, CuContext(0x00000000026b2bc0, instance fe73669339970426))))

ERROR: a exception was thrown during kernel execution.
Run Julia on debug level 2 for device stack traces.
```

The message is lost, of course, but we might be able to properly report it in the future.

Reply:
We need a bit more complicated code here, right?

```julia
julia> kernel(x) = (x && throw(ArgumentError("pre $x post")); nothing)
kernel (generic function with 1 method)

julia> CUDA.@sync @cuda kernel(false);
ERROR: InvalidIRError: compiling kernel kernel(Bool) resulted in invalid LLVM IR
Reason: unsupported dynamic function invocation (call to print_to_string(xs...) in Base at strings/io.jl:124)
```

Reply:
Ah yes the GPUCompiler hack doesn't properly detect GC frames, so it stumbles on the `ArgumentError(...)` here.